supernova: enable building JACK API on macOS

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

Currently supernova is hardcoded to use PortAudio on macOS. This PR modifies CMakeLists file to mimic the API choice in scsynth (with the exception of choosing portaudio instead of coreaudio by default on macOS).

## Types of changes

<!-- Delete lines that don't apply -->

- Documentation
- Bug fix (?)
- New feature (?)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested (except for Windows, as supernova doesn't build on Windows)
- [x] scsynth and supernova boot from the terminal using JACK backend
- [x] scsynth and supernova boot from sclang using JACK backend
- [x] supernova's CMakeLists logic is updated to mimic scsynth's CMakeLists
- [x] Updated documentation
- [x] Commit history is cleaned up and consolidated
- [x] This PR is ready for review

## Comments

Previously I had a problem connecting to JACK server when the scsynth/supernova were booted from SC: 
<details><summary>(details)</summary>
<p>

~~I am able to start servers from command line, and they connect to JACK backend and make sound. However, if I try booting server from SC, I'm getting~~

```
Booting server 'localhost' on address 127.0.0.1:57110.
Found 0 LADSPA plugins
Faust: supercollider.cpp: sc_api_version = 3
Faust: FaustJPverbRaw numControls=11
Faust: supercollider.cpp: sc_api_version = 3
Faust: FaustGreyholeRaw numControls=7
could not initialize audio.
Server 'localhost' exited with exit code 0.
```
for scsynth
and
```
Booting server 'localhost' on address 127.0.0.1:57110.
Supernova booting
Found 0 LADSPA plugins
Server 'localhost' exited with exit code 0.
```
for supernova.

~~Again, if I boot the server in terminal with the same options (taken from s.options.asOptionsString), it starts and works. I'd appreciate ideas how to debug this further...~~

</p>
</details>

I realized, that this was due to JACK (installed by homebrew in `/usr/local/bin`) not being in the `$PATH` in the SC environment. 

Solution:

```supercollider
"PATH".setenv("echo $PATH".unixCmdGetStdOut ++ ":/usr/local/bin"); //add to $PATH
```
And also make the scsynth/supernova automatically connect to system inputs/outputs:
```supercollider
// connect all input channels with system
"SC_JACK_DEFAULT_INPUTS".setenv("system");
// connect all output channels with system
"SC_JACK_DEFAULT_OUTPUTS".setenv("system");
```

After that, I was able to start scsynth/supernova and connect it to (previously started) jack server.

Caveats:
- `jackd` seems to require a single device for input and output, so in order to use it with the internal soundcard on a mac, one needs to create an aggregate device
- I have an intermittent problem with the inputs (in jack itself) not having any signal present; can't find a reliable reproducer though, this seems to be an issue with the jack itself and there seems to be an open issue about it: https://github.com/jackaudio/jack1/issues/65

I was able to start jack either from terminal, `qjackctl` gui, or with SC:

```supercollider
~jackPid = "jackd -d coreaudio -r48000 -p512".unixCmd;
```

JACK was installed with
```bash
brew install jack
```
and
```bash
brew install qjackctl
```
for the GUI allowing control over JACK connections.


For reference: what sparked this PR is a [problem I discovered with JackOSX](https://github.com/jackaudio/jack2/issues/424).

edit:
I tested this with both jack1 from homebrew and jack2 from JackOSX.0.92_b3. Compilation works in both cases (with some problem in the verification stage for jack2).

The takeaway is that both JACKs seem to be a little _broken_:
- jack1 has intermittent problems with audio in
- jack2 with JackRouter does not clean up resources when exiting, so the `coreaudiod` process is left with ~150MB of garbage memory (?) after each start/stop of jack server

mtmccrea commented:
> [ ] Updated documentation (does this need documentation?)

All of your observations about installation, caveats, and starting jack by various means seem worth documenting. I could see this in either a new section of `README_MACOS.md` or its own "Jack on OSC" wiki or `.md`

dyfer commented:
I added documentation for building and running SC with JACK on macOS.

For the sake of completeness, I also tested with JackOSX (which is jack2, as opposed to jack1 from homebrew). The build process _mostly_ worked, SC starts and I can make sound using SC<->JACK, however, when building, the `verify_app` stage failed:
```
-- Verifying app
-- ===========================================================================
-- Analyzing app='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app'
-- bundle='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app'
-- executable='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider'
-- valid='1'
-- executable file 1: /somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Frameworks/QtWebEngineCore.framework/Versions/5/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
-- executable file 2: /somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/sclang
-- executable file 3: /somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider
-- executable file 4: /somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/scsynth
-- executable file 5: /somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/supernova
-- verified='0'
-- info='external prerequisites found:
f='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/MacOS/sclang'
external_prereqs='/Library/Frameworks/Jackmp.framework/Versions/A/Jackmp'
;external prerequisites found:
f='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/scsynth'
external_prereqs='/Library/Frameworks/Jackmp.framework/Versions/A/Jackmp'
;external prerequisites found:
f='/somePath/src/supercollider/build/Install/SuperCollider/SuperCollider.app/Contents/Resources/supernova'
external_prereqs='/Library/Frameworks/Jackmp.framework/Versions/A/Jackmp'
'
--
CMake Error at /usr/local/Cellar/cmake/3.13.3/share/cmake/Modules/BundleUtilities.cmake:1137 (message):
  error: verify_app failed
Call Stack (most recent call first):
  editors/sc-ide/cmake_install.cmake:96 (verify_app)
  editors/cmake_install.cmake:37 (include)
  cmake_install.cmake:55 (include)


make: *** [install_buildpart_0] Error 1

** BUILD FAILED **


The following build commands failed:
	PhaseScriptExecution CMake\ PostBuild\ Rules build/SuperCollider.build/RelWithDebInfo/install.build/Script-89009FFBE94745E286AE1E35.sh
(1 failure)
```
Jack libraries are present (symlinked) at `/Library/Frameworks/Jackmp.framework/Versions/A/Jackmp`. I don't understand libraries linking (...) so I'm not sure if this is useful, but `supernova` (as well as `scsynth`) do point to that library:

```
myName$ otool -L Install/SuperCollider/SuperCollider.app/Contents/Resources/scsynth
Install/SuperCollider/SuperCollider.app/Contents/Resources/scsynth:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
	@loader_path/../Frameworks/libsndfile.1.dylib (compatibility version 2.0.0, current version 2.28.0)
	/Library/Frameworks/Jackmp.framework/Versions/A/Jackmp (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 934.0.0)
	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation (compatibility version 300.0.0, current version 1560.12.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/System/Library/Frameworks/CFNetwork.framework/Versions/A/CFNetwork (compatibility version 1.0.0, current version 975.0.3)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1560.12.0)
	/usr/lib/libobjc.A.dylib (compatibility version 1.0.0, current version 228.0.0)
```
I'd appreciate any pointers how to address the `verify_app` issue, thanks!

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+if(AUDIOAPI STREQUAL "coreaudio")
+	set(AUDIOAPI portaudio)
+  message(STATUS "Supernova: coreaudio API not implemented, using ${AUDIOAPI} instead")
+endif()
```
indent all new cmake code with 4 spaces please

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+
+if(AUDIOAPI STREQUAL "coreaudio")
+	set(AUDIOAPI portaudio)
+  message(STATUS "Supernova: coreaudio API not implemented, using ${AUDIOAPI} instead")
```
if it's not supported, i think this should be a fatal error rather than defaulting to something the user may not want. perhaps we need a separate CMake option for supernova audio api?

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+		set(AUDIOAPI portaudio)
+	else()
+		set(AUDIOAPI jack)
+	endif(APPLE)
```
can be simplified to

```cmake
if(AUDIOAPI STREQUAL "default")
    if(APPLE OR WIN32)
		set(AUDIOAPI portaudio)
	else()
		set(AUDIOAPI jack)
	endif()
endif()
```

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+    find_package(Jack)
+    if(NOT JACK_FOUND)
+        message(FATAL_ERROR "Jack selected as audio API, but development files not found")
+    endif()
```
`find_package(Jack REQUIRED)` will accomplish the same thing more effectively. you could also update the portaudio check below the same way

https://cmake.org/cmake/help/v3.5/command/find_package.html

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
     endif()
-else()
+endif()
+message(STATUS "Supernova audio API: ${AUDIOAPI}")
```
don't capitalize supernova

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+message(STATUS "Supernova audio API: ${AUDIOAPI}")
+
+if (AUDIOAPI STREQUAL jack)
     find_library(JACK NAMES jack)
```
is this necessary with `find_package` above? seems like the same work done twice in different ways. also dubious of just failing silently if jack isn't found here

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+endif()
+message(STATUS "Supernova audio API: ${AUDIOAPI}")
+
+if (AUDIOAPI STREQUAL jack)
```
change this and the portaudio check below from `jack` to `"jack"` -- otherwise, cmake will first see if a variable with the name `jack` exists, which is not what you want

dyfer replied:
this is taken from the scsynth CMakeLists:
https://github.com/supercollider/supercollider/blob/ebe77610fd745d62ba5c36780d42f5da54d35362/server/scsynth/CMakeLists.txt#L18
should this be changed in scsynth's CMakeLists as well?

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
+elseif (AUDIOAPI STREQUAL portaudio)
+    target_include_directories(libsupernova PUBLIC ${PORTAUDIO_INCLUDE_DIRS})
+    if (WIN32)
+        target_compile_definitions(libsupernova PUBLIC PORTAUDIO_BACKEND -DPORTAUDIO_BACKEND)
```
`-DPORTAUDIO_BACKEND` is superfluous here -- `-D` is added to the beginning of compile definitions automatically

dyfer replied:
This is exactly what Windows section had previously, and I can't currently compile supernova on Windows so can't test such change. Is it still worth changing?

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -104,31 +104,57 @@ else()
     endif()
+elseif (AUDIOAPI STREQUAL portaudio)
+    target_include_directories(libsupernova PUBLIC ${PORTAUDIO_INCLUDE_DIRS})
+    if (WIN32)
```
remove spaces between `if`/`elseif` and paren throughout the code you added

dyfer commented on README_MACOS.md:
````diff
@@ -295,6 +295,58 @@ This should be fixed at some point (its a build tool configuration issue). Until
+In order to build with JACK, you need to add the `-DAUDIOAPI=jack` flag to cmake. The full command to configure the build (with supernova) would be:
+
+```
+cmake -G Xcode -DCMAKE_PREFIX_PATH=`brew --prefix qt5` -DSUPERNOVA=ON -DAUDIOAPI=jack ..
````
I included this since I personally like full commands I can just copy/paste - I know this is lazy, but I like to have that option. If you insist this is suboptimal, I'm happy to remove that. 

My question is in what case on macOS the command would be different? If you have a custom place for QT you probably know enough to change it, and if not, this is helpful for newbies (which is what I was for quite some time as far as building goes)

brianlheim replied:
there are over a dozen options you can pass to cmake to configure your supercollider build, some of them are documented in this very file.

i understand the desire to be beginner friendly but this really isn't that in my opinion, it's just creating clutter. let's teach people how to effectively use the build system if they don't know, rather than giving complete commands for every possible build scenario. if this command were universal or even the common case, you might have a point.

dyfer replied:
> if this command were universal or even the common case, you might have a point.

IMO this was the common case for this particular scenario (building with JACK) but I see your point it might not be so I'll remove it.

dyfer commented on README_MACOS.md:
```diff
@@ -295,6 +295,58 @@ This should be fixed at some point (its a build tool configuration issue). Until
+
+##### Running SuperCollider with JACK
+
+When `jack` is installed via `homebrew`, you need to add jack's path to the `$PATH` environment variable in `sclang`:
```
> for instance, could we tell the servers via an env var or command line arg where the JACK executable is?

I'm not aware how to do this differently...

> same with setting SC_JACK_DEFAULT_INPUTS below. all of this could be made simpler and wrapped inside a helper function on ServerOptions. my question there is, do people want to use SC_JACK_DEFAULT_INPUTS=system by default? what do they gain by doing so?

`SC_JACK_DEFAULTS_INPUTS` is _I think_ the way you would do that on linux anyway... AFAIR this connects SC inputs/outputs to the soundcard. Otherwise, SC starts unconnected, so no sound is coming to the speakers (only useful if you are manually routing it to another app).

brianlheim replied:
i am asking you to try to think of a different way of doing it so people aren't manually managing environment variables.

dyfer replied:
Managing of JACK connections is explained this way in the [help](https://github.com/supercollider/supercollider/blob/3e44dfe65d39a3e180f0a4203fc7b2cd3c8bfc35/HelpSource/Reference/AudioDeviceSelection.schelp#L90). Do you think I should come up with a different way to do this? If I do, then this Readme is probably not the best place to do it. My point reiterating info from the help here is to get user up and running quicker, but I'd be also fine pointing them to the help.

For adding jack executable to the $PATH, now I need to go back and revisit why did I have to do it in the first place.

dyfer commented on README_MACOS.md:
```diff
@@ -295,6 +295,58 @@ This should be fixed at some point (its a build tool configuration issue). Until
+
+Afterwards proceed with the build process as usual.
+
+##### Running SuperCollider with JACK
```
`#####` was intentionalas it's used elsewhere in this file as well, but if you prefer to have `###` here that's fine.

dyfer commented:
@brianlheim I can't seem to respond to the comment pointing to code that has changed (?), here's more response regarding switching backend from coreaudio to portaudio in supernova:
>>if it's not supported, i think this should be a fatal error rather than defaulting to something the user may not want. perhaps we need a separate CMake option for supernova audio api?

I personally treat it as a temporary situation where we don't have coreaudio backend for supernova. If you feel strongly about it, I can change it to throw an error if coreaudio is explicitly requested. I think adding more granularity to cmake is not optimal, there's already a lot there... but if you think that's best then I can add that. I personally don't think having 2 servers with different backends is useful (unless necessary, as it is now on macOS), but again I'm happy to be persuaded about it.

dyfer commented:
I rebased this PR to include the latest changes from `develop`. 

I also added an error message if the user explicitly requests `coreaudio` backend. Currently I don't think that adding separate audio api switches for the two servers is beneficial (ideally we should have all APIs implemented in both). 
This is the current audio API selection logic on macOS:

| requested API | scsynth | supernova |
| - | - | - |
| default | coreaudio | portaudio |
| coreaudio | coreaudio | ERROR |
| portaudio | portaudio | portaudio |
